hbase-server/src/main/java/org/apache/hadoop/hbase/master/HMaster.java

```diff
@@ -4155,18 +4155,25 @@ public CompactionState getCompactionState(final TableName tableName) {
           continue;
         }
         RegionMetrics regionMetrics = sl.getRegionMetrics().get(regionInfo.getRegionName());
-        if (regionMetrics.getCompactionState() == CompactionState.MAJOR) {
-          if (compactionState == CompactionState.MINOR) {
-            compactionState = CompactionState.MAJOR_AND_MINOR;
-          } else {
-            compactionState = CompactionState.MAJOR;
-          }
-        } else if (regionMetrics.getCompactionState() == CompactionState.MINOR) {
-          if (compactionState == CompactionState.MAJOR) {
-            compactionState = CompactionState.MAJOR_AND_MINOR;
-          } else {
-            compactionState = CompactionState.MINOR;
+        if (regionMetrics != null) {
+          if (regionMetrics.getCompactionState() == CompactionState.MAJOR) {
+            if (compactionState == CompactionState.MINOR) {
+              compactionState = CompactionState.MAJOR_AND_MINOR;
+            } else {
+              compactionState = CompactionState.MAJOR;
+            }
+          } else if (regionMetrics.getCompactionState() == CompactionState.MINOR) {
+            if (compactionState == CompactionState.MAJOR) {
+              compactionState = CompactionState.MAJOR_AND_MINOR;
+            } else {
+              compactionState = CompactionState.MINOR;
+            }
           }
+        } else {
+          String getCompactionDetailsErrorMsg = String.format("Can not get compaction details for region: %s, it may be disabled or in transition.",
```

Comment:
I think we should just skip the region if it is not online? FWIW, if a region is not online, it can not be compacting right?

Reply:
The other regions of the table still can compact when one of them is not online.

This is a another issue which i am thinking about.
the compact state of the table should be minor/major compaction if a region is not online and others is compacting. 
however, in the case, we get unkown comapct state about the table from HBase UI (table details page)

I think only in this case that the table has closed regions and other regions compact state is NONE，the compact state of table is UNKOWN.

Reply:
I just mean you should skip check the state of this region, instead of throwing an exception which fails the whole operation. It will not affect the result if we just do not count this region. If other regions are compacting, the compaction state will be MINOR, MAJOR or MINOR and MAJOR, it will not be UNKNOWN, unless all regions are not online.

Reply:
> I just mean you should skip check the state of this region, instead of throwing an exception which fails the whole operation. 

OK, I will update it later

Reply:
> If other regions are compacting, the compaction state will be MINOR, MAJOR or MINOR and MAJOR, it will not be UNKNOWN, unless all regions are not online.

I still have question in here.
If both notOnline and online regions exist and online regions is not compacting (NONE), what state should be returned ?  I think it is UNKOWN, what about you?

Reply:
I think it should be NONE. As if a region is not online, it can not be compacting right? A compaction must be executed on a region server, which means the region should be online.

Reply:
yes, thanks